File: PathoFroh/src/main/java/com/patho/main/repository/jpa/custom/impl/ListItemRepositoryImpl.java
package com.patho.main.repository.jpa.custom.impl;

import com.patho.main.model.preset.ListItem;
import com.patho.main.model.preset.ListItemType;
import com.patho.main.model.preset.ListItem_;
import com.patho.main.repository.jpa.custom.ListItemRepositoryCustom;

import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.util.ArrayList;
import java.util.List;

public class ListItemRepositoryImpl extends AbstractRepositoryCustom implements ListItemRepositoryCustom {

    @Override
    public List<ListItem> findAll(ListItemType listType, boolean irgnoreArchived) {
        CriteriaQuery<ListItem> criteria = getCriteriaBuilder().createQuery(ListItem.class);
        Root<ListItem> root = criteria.from(ListItem.class);

        criteria.select(root);

        List<Predicate> predicates = new ArrayList<Predicate>();
        predicates.add(getCriteriaBuilder().equal(root.get(ListItem_.listType), listType));

        if (irgnoreArchived)
            predicates.add(getCriteriaBuilder().equal(root.get(ListItem_.archived), false));

        criteria.where(getCriteriaBuilder().and(predicates.toArray(new Predicate[predicates.size()])));

        criteria.distinct(true);

        return getSession().createQuery(criteria).getResultList();
    }

    public List<ListItem> findAllOrderByIndex(ListItemType listType, boolean irgnoreArchived) {

        CriteriaQuery<ListItem> criteria = getCriteriaBuilder().createQuery(ListItem.class);
        Root<ListItem> root = criteria.from(ListItem.class);

        criteria.select(root);

        List<Predicate> predicates = new ArrayList<Predicate>();
        predicates.add(getCriteriaBuilder().equal(root.get(ListItem_.listType), listType));

        if (irgnoreArchived)
            predicates.add(getCriteriaBuilder().equal(root.get(ListItem_.archived), false));

        criteria.where(getCriteriaBuilder().and(predicates.toArray(new Predicate[predicates.size()])));

        criteria.orderBy(getCriteriaBuilder().asc(root.get(ListItem_.indexInList)));

        criteria.distinct(true);

        return getSession().createQuery(criteria).getResultList();

    }

}

File: PathoFroh/src/main/kotlin/com/patho/main/repository/jpa/custom/impl/ListItemRepositoryImpl.kt
package com.patho.main.repository.jpa.custom.impl

import com.patho.main.model.preset.ListItem
import com.patho.main.model.preset.ListItemType
import com.patho.main.model.preset.ListItem_
import com.patho.main.repository.jpa.custom.ListItemRepositoryCustom
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*
import javax.persistence.criteria.Predicate

@Service
open class ListItemRepositoryImpl : AbstractRepositoryCustom(), ListItemRepositoryCustom {

    @Transactional
    override fun findAll(listType: ListItemType, ignoreArchived: Boolean): List<ListItem> {
        val criteria = criteriaBuilder.createQuery(ListItem::class.java)
        val root = criteria.from(ListItem::class.java)
        criteria.select(root)
        val predicates: MutableList<Predicate> = ArrayList()
        predicates.add(criteriaBuilder.equal(root.get(ListItem_.listType), listType))
        if (ignoreArchived) predicates.add(criteriaBuilder.equal(root.get(ListItem_.archived), false))
        criteria.where(criteriaBuilder.and(*predicates.toTypedArray()))
        criteria.distinct(true)
        return session.createQuery(criteria).resultList
    }

    @Transactional
    override fun findAllOrderByIndex(listType: ListItemType, ignoreArchived: Boolean): List<ListItem> {
        val criteria = criteriaBuilder.createQuery(ListItem::class.java)
        val root = criteria.from(ListItem::class.java)
        criteria.select(root)
        val predicates: MutableList<Predicate> = ArrayList()
        predicates.add(criteriaBuilder.equal(root.get(ListItem_.listType), listType))
        if (ignoreArchived) predicates.add(criteriaBuilder.equal(root.get(ListItem_.archived), false))
        criteria.where(criteriaBuilder.and(*predicates.toTypedArray()))
        criteria.orderBy(criteriaBuilder.asc(root.get(ListItem_.indexInList)))
        criteria.distinct(true)
        return session.createQuery(criteria).resultList
    }
}